cli/internal/client/client.go

```diff
@@ -143,11 +145,28 @@ func (c *ApiClient) PutArtifact(hash string, duration int, rawBody interface{})
 	if encoded != "" {
 		encoded = "?" + encoded
 	}
-	req, err := retryablehttp.NewRequest(http.MethodPut, c.makeUrl("/v8/artifacts/"+hash+encoded), rawBody)
+	// Read the entire artifactReader into memory so we can easily compute the Content-MD5.
+	// Note: retryablehttp.NewRequest reads the artifactReader into memory so there's no
+	// additional overhead by doing the ioutil.ReadAll here instead.
+	artifactBody, err := ioutil.ReadAll(artifactReader)
+	if err != nil {
+		return fmt.Errorf("failed to store files in HTTP cache: %w", err)
+	}
+	md5Sum := md5.Sum(artifactBody)
+	contentMd5 := base64.StdEncoding.EncodeToString(md5Sum[:])
+
+	req, err := retryablehttp.NewRequest(http.MethodPut, c.makeUrl("/v8/artifacts/"+hash+encoded), artifactBody)
+	Trailer := make(http.Header)
+	Trailer.Add("Content-MD5", contentMd5)
+	// We need to initialize the trailer since it's not auto-initialized by the http module
+	req.Trailer = Trailer
+	// ContentLenth -1 is requried to set trailers on the http request.
+	req.ContentLength = -1
```

Comment:
This might be the only unexpected api change. But this is required, as otherwise the Trailer would not be set by the http module https://pkg.go.dev/net/http#:~:text=//%20For%20client%20requests%2C%20Trailer,must%0A%09//%20not%20mutate%20Trailer.%0A%09//

Reply:
The resulting headers are:

```
PUT /v8/artifacts/hash?slug=my-team-slug HTTP/1.1
Host: localhost:8888
Transfer-Encoding: chunked
Accept-Encoding: gzip
Authorization: Bearer my-token
Content-Type: application/octet-stream
User-Agent: turbo v1 go1.17.8 darwin (arm64)
X-Artifact-Duration: 500

12
My string artifact
0

```

Reply:
Should the trailer show up here? And I think you maybe need to set a header `Trailer: Content-MD5`?

Reply:
I think it should be showing. Looking into why it's not. It might be the way DumpRequest works. 
Good catch I added the "Trailer: Content-MD5" header.

Reply:
the request is definitely sending the Trailer itself, but hiding it in DumpRequest.

Reply:
Weird. Ok, it looks like it is definitely coming through though if we can read it in the handler.

Reply:
Yea. I also looked in the debugger when sending the request, it's there. So it might be a config the test server as to why the header isn't present.